### DISCUSSION TITLE
Continuous repeated CIDRNotAvailable event for one node in cluster.

### DISCUSSION THREAD

**Discussion by tjf** [k3s-io/k3s#7663] [Q&A] [answered]
**Environmental Info:**
K3s Version: 

```
k3s version v1.26.4+k3s1 (8d0255af)
go version go1.19.8
```

Node(s) CPU architecture, OS, and Version: 

`Linux node1 5.10.0-16-amd64 #1 SMP Debian 5.10.127-1 (2022-06-30) x86_64 GNU/Linux`

Cluster Configuration:

Total of 3 servers using the embedded etcd.  I am using Calico 3.25 instead of Flannel.

**Describe the bug:**

While the node is working fine -- all network functionality, pod scheduling, service routing, etc -- it is constantly logging CIDRNotAvailable events. I can't find any mention in the k3s logs of any network issues that could be causing this.  If I didn't see this in the `kubectl describe node` output, I'd have no idea there was a problem, honestly.
```
Events:
  Type    Reason            Age                      From           Message
  ----    ------            ----                     ----           -------
  Normal  CIDRNotAvailable  107s (x1025 over 3d14h)  cidrAllocator  Node node1 status is now: CIDRNotAvailable
```

**Steps To Reproduce:**

Not sure, but if someone could point me in the direction of troubleshooting further, I'd appreciate it.

**Expected behavior:**

I expect this event not to be logged :-)

**Additional info:**
Here's the systemd unit file ExecStart entry, it's the same on all the nodes:

```
ExecStart=/usr/local/bin/k3s \
    server \
	'--disable-network-policy' \
	'--flannel-backend=none' \
	'--token=<redacted>' \
	'--disable=servicelb' \
	'--cluster-cidr=192.168.20.0/23' \
	'--cluster-dns=172.20.0.10' \
	'--service-cidr=172.20.0.0/24' \
	'--server' \
	'https://192.168.45.52:6443' \
```

Also -- I opened an issue with Project Calico, but they don't think this is being caused by the calico software.  See https://github.com/projectcalico/calico/issues/7722

Thanks :-)

**Comment by brandond** [comment, marked as answer]
```
	'--cluster-cidr=192.168.20.0/23' \
	'--cluster-dns=172.20.0.10' \
	'--service-cidr=172.20.0.0/24' \
```

Why the tiny CIDRs? The default node CIDR mask is /24, so with a /23 cluster-cidr you only have enough addresses for two nodes in your cluster. You've also only allocated yourself enough space for 255 ClusterIP services.

You should rebuild your cluster with more reasonable CIDR blocks.

**Reply by tjf** [reply]
> The default node CIDR mask is /24

AHA!.. This is the missing piece of the puzzle.  I was not aware of this.  Thank you.

However, I'm using Calico, and not Flannel.  As a result, it ignores the podCIDR value anyway and is splitting that /23 up into a bunch of /26's and spreading them out.  Everything is working fine, so I'm guessing I have the kube-controller-manager AND calico thinking they're in charge of IPAM which is why I'm getting the error (it can't split the /23 into 3 /24's).

Re: tiny CIDRs, I guess I'm just an old school network engineer used to having to justify every lick of space I need.  😆  ... For my use case, these are way more than reasonable.

Thank you!

**Reply by brandond** [reply]
yeah, you could probably disable the ipam controller; something like `--kube-controller-manager-arg=controllers=*,-nodeipam` might do it but I haven't tested that.

**Reply by tjf** [reply]
I just gave this a shot, but it didn't appear to make any difference.  I'm guessing the implicit `--allocate-node-cidrs` to the kube-controller-manager process is keeping it alive.  Is there a way I can negate that?

**Reply by brandond** [reply]
You could try `--kube-controller-manager-arg=allocate-node-cidrs=false` ?

**Reply by tjf** [reply]
🤦🏻‍♂️ .. yea, that would make sense.  LOL

That seems to have corrected the issue.  No longer seeing CIDRNotAvailable on my node's events.  I'll update the remainder of nodes to include that as well.  

Thanks again for your time and work on this great project.  🙏🏼

Ninja Edit: No!.. it's not corrected.  :-( ... I see there's anotehr animal in the logs I may need to adjust as well -- `cloud-controller-manager`

```
Jun 01 20:23:52 node1 k3s[2054467]: time="2023-06-01T20:23:52-04:00" level=info msg="Running kube-controller-manager --allocate-node-cidrs=false --authentication-kubeconfig=/var/lib...
Jun 01 20:23:52 node1 k3s[2054467]: time="2023-06-01T20:23:52-04:00" level=info msg="Running cloud-controller-manager --allocate-node-cidrs=true --authentication-kubeconfig=/var/lib...
```

I'll keep messing with it.  Thx again!

**Reply by tjf** [reply]
Just adding another comment in the event you missed my edit above.

**Reply by brandond** [reply]
The k3s stub ccm doesn't handle node cidrs, so I think you should be ok.

**Reply by tjf** [reply]
Adding `--kube-cloud-controller-manager-arg=allocate-node-cidrs=false` to the command line seems to have squashed it!

Update: Bah, no it didn't.  I'll dig into the code and see if I can hunt this down.

**Reply by BhautikChudasama** [reply]
@tjf how did you fixed this issue?

**Reply by BhautikChudasama** [reply]
@tjf This is the part of `--kube-controller-manager-arg` not `--kube-cloud-controller-manager-arg`
**Full example**
```
--kube-controller-manager-arg=allocate-node-cidrs=false
```

**Comment by tjf** [comment]
@BhautikChudasama thanks for replying!  I never did end up fixing this as it was just on a personal cluster and it didn't matter THAT much.  But, seeing how I messed up the args, I may give this another shot.  Good job!
